README.md

Stream realtime low-latency h264 video directly to the browser.
Comes with a browser player, and streaming server.
 Uses [Broadway](https://github.com/mbebenita/Broadway) browser decoder and player. Ideal for for Raspberry pi cam streaming.

# Usage

## Server:

 ```js
 const AvcServer = require('ws-avc-player/lib/server')
 const { WebSocketServer } = require('@clusterws/cws') // works with ws, legacy uws
 const wss = new WebSocketServer({ port: 3333 })
 const avcServer = new AvcServer(wss, 640, 480) //initial width and height (it adapts to the stream)

 avcServer.setVideoStream(h264Stream)
 ```
More detailed in [example/index.js](example/index.js) 

## Client: 

```html
<html>
  <body>
    <!-- define the element to hold the canvas -->
     <div id="video-box" />
    <!-- provide WSAvcPlayer -->
    <script type="text/javascript" src="WSAvcPlayer.js" />
    <script type="text/javascript">
      //initialize the player, if useWorker: true, than you must have `/Decoder.js` availible at the root of the domain.
      var wsavc = new WSAvcPlayer.default({useWorker:false}); 
      //append the canvas to the box element, you can style the box element and canvas.
      document.getElementById('video-box').appendChild(wsavc.AvcPlayer.canvas)
      //connect to the websocket
      wsavc.connect("ws://" + document.location.host+":3333");
    </script>    
  </body>
</html>
```

More detailed in [example/html/index.html](example/html/index.html)

# Running the demo
```bash
git clone https://github.com/matijagaspar/ws-avc-player
cd ws-avc-player
npm install
npm run example

# browse to http://127.0.0.1:8080/ for a demo player

```
then run

```bash
ffmpeg -framerate 30 -video_size 640x480 -f [driver] -i [device]  -vcodec libx264 -vprofile baseline -b:v 500k -bufsize 600k -tune zerolatency -pix_fmt yuv420p -r 15 -g 30 -f rawvideo tcp://localhost:5000
```

or
```bash
raspivid -t 0 -w 640 -h 480 -hf -fps 15 -o - | nc localhost 5000
```

alternatively run:
```bash
npm run example raspivid
```
It will automatically run raspivid too

## Using it in your own project

`yarn add ws-avc-player`

### On Client
* `import WSAvcPlayer from 'ws-avc-player'`
* ``` 
        const wsavc = new WSAvcPlayer({useWorker:true})
        wsavc.connect(serverUrl);

  ```
###  On Server:

* See `example/index.js`




# TODO:
 * ~Decoder as worker~
 * More docs
 * More cleanup
 * Audio
 * ~Ability to change video resolution or better parse sps/pps~
 * ~Decent performance~
 
# Credits
* [matijagaspar](https://github.com/matijagaspar)
* [h264-live-player](https://github.com/131/h264-live-player)
* [Broadway](https://github.com/mbebenita/Broadway)
* [urbenlegend/WebStreamer](https://github.com/urbenlegend/WebStreamer)


CHANGELOG.md
## 2.0.2 (2019-10-21)
* Update readme example

## 2.0.1 (2019-10-21)
* small readme update

## 2.0.0 (2019-10-21)

* feat: embed wasm into the lib so it works out of the box ([86ab22d](https://github.com/matijagaspar/ws-avc-player/commit/86ab22d))
* add Broadway as submoduler, cleanup ([22b966a](https://github.com/matijagaspar/ws-avc-player/commit/22b966a))
* make use of broadway player ([bfb9e90](https://github.com/matijagaspar/ws-avc-player/commit/bfb9e90))
* update dependencies, improve example, get rid of player canvas styling ([e2b43ce](https://github.com/matijagaspar/ws-avc-player/commit/e2b43ce))
* Update README.md ([e35b27a](https://github.com/matijagaspar/ws-avc-player/commit/e35b27a))





example/index.js
/* eslint-disable no-console */
const AvcServer = require('../lib/server')
const path = require('path')
const http = require('http')
// const WebSocketServer = require('uws').Server
const { WebSocketServer } = require('@clusterws/cws')
// require('uWebSockets.js')
const net = require('net')
const spawn = require('child_process').spawn

const useRaspivid = process.argv.includes('raspivid')
const width = 1280
const height = 720

const express = require('express')
const app = express()
// serve the html/index.html
app.use(express.static(path.resolve(__dirname, 'html')))
// serve the player
app.use(express.static(path.resolve(__dirname, '../lib')))

const server = http.createServer(app)

// init web socket
const wss = new WebSocketServer({ /* port: 3333 */ server })
// init the avc server.
const avcServer = new AvcServer(wss, width, height)

// handling custom events from client
avcServer.client_events.on('custom_event_from_client', e => {
    console.log('a client sent', e)
    // broadcasting custom events to all clients (if you wish to send a event to specific client, handle sockets and new connections yourself)
    avcServer.broadcast('custom_event_from_server', { hello: 'from server' })
})

// RPI example
if (useRaspivid) {
    let streamer = null

    const startStreamer = () => {
        console.log('starting raspivid')
        streamer = spawn('raspivid', [ '-pf', 'baseline', '-ih', '-t', '0', '-w', width, '-h', height, '-hf', '-fps', '15', '-g', '30', '-o', '-' ])
        streamer.on('close', () => {
            streamer = null
        })
        avcServer.setVideoStream(streamer.stdout)
    }

    // OPTIONAL: start on connect
    avcServer.on('client_connected', () => {
        if (!streamer) {
            startStreamer()
        }
    })


    // OPTIONAL: stop on disconnect
    avcServer.on('client_disconnected', () => {
        console.log('client disconnected')
        if (avcServer.clients.size < 1) {
            if (!streamer) {
                console.log('raspivid not running')
                return
            }
            console.log('stopping raspivid')
            streamer.kill('SIGTERM')
        }
    })

} else {
// create the tcp sever that accepts a h264 stream and broadcasts it back to the clients
    this.tcpServer = net.createServer((socket) => {
    // set video stream
        socket.on('error', e => {
            console.log('video downstream error:', e)
        })
        avcServer.setVideoStream(socket)

    })
    this.tcpServer.listen(5000, '0.0.0.0')
}

server.listen(8081)

// if not using raspivid option than use one of this to stream
// ffmpeg OSX
// then run ffmpeg: ffmpeg -framerate 30 -video_size 640x480 -f avfoundation -i 0  -vcodec libx264 -vprofile baseline -b:v 500k -bufsize 600k -tune zerolatency -pix_fmt yuv420p -r 15 -g 30 -f rawvideo tcp://localhost:5000

// fmpeg Windows:

// ffmpeg -framerate 25 -video_size 640x480 -f dshow -i "video=<DEVICE>"  -vcodec libx264 -vprofile baseline -b:v 500k -bufsize 600k -tune zerolatency -pix_fmt yuv420p -f rawvideo tcp://localhost:5000

// to get video devices run:
// ffmpeg -list_devices true -f dshow -i dummy

// To get options of the device: 
// ffmpeg -f dshow -list_options true -i video="<Device>"


// examples:
// ffmpeg -framerate 25 -video_size 640x480 -f dshow -i video="HD Camera"  -vcodec libx264 -vprofile baseline -b:v 500k -bufsize 600k -tune zerolatency -pix_fmt yuv420p -f rawvideo tcp://localhost:5000

// ffmpeg -framerate 25 -video_size 1280x720 -f dshow -i "video=Logitech HD Webcam C270"  -vcodec libx264 -vprofile baseline -b:v 500k -bufsize 600k -tune zerolatency -pix_fmt yuv420p -f rawvideo tcp://localhost:5000
// ffmpeg.exe -framerate 30 -video_size 1280x720 -f dshow -i video="HD Camera"  -vcodec libx264 -vprofile baseline -b:v 2m -bufsize 2m -pass 1 -coder 0 -bf 0 -flags -loop -tune zerolatency -pix_fmt yuv420p -wpredp 0 -f rawvideo tcp://localhost:5000
// RPI
// /opt/vc/bin/raspivid -pf baseline -ih -t 0 -w 640 -h 480 -hf -fps 15 -g 30 -o - | nc localhost 5000
